interop_binaries/tests/end_to_end.rs

```diff
@@ -0,0 +1,601 @@
+        CONTAINER_CLIENT.run(RunnableImage::from(Postgres::default()).with_tag("14-alpine"));
+    let helper_postgres_port = helper_db_container.get_host_port_ipv4(5432);
+
+    // Pick four ports for HTTP servers.
+    let client_port = pick_unused_port().context("couldn't pick a port for the client")?;
+    let leader_port = pick_unused_port().context("couldn't pick a port for the leader")?;
+    let helper_port = pick_unused_port().context("couldn't pick a port for the helper")?;
+    let collector_port = pick_unused_port().context("couldn't pick a port for the collector")?;
+    assert_eq!(
+        BTreeSet::from([client_port, leader_port, helper_port, collector_port]).len(),
+        4,
+        "Ports selected for HTTP servers were not unique",
+    );
+
+    // Step 1: Create and start containers. (here, we just run the binaries instead)
+    // We use std::process instead of tokio::process so that we can kill the child processes from
+    // a Drop implementation. tokio::process::Child::kill() is async, and could not be called from
+    // there.
+    let mut client_command = Command::new(env!("CARGO_BIN_EXE_janus_interop_client"));
+    client_command.arg("--port").arg(format!("{}", client_port));
+    let mut leader_command = Command::new(env!("CARGO_BIN_EXE_janus_interop_aggregator"));
+    leader_command.arg("--port").arg(format!("{}", leader_port));
+    leader_command.arg("--postgres-url").arg(format!(
+        "postgres://postgres@127.0.0.1:{}/postgres",
+        leader_postgres_port
+    ));
+    let mut helper_command = Command::new(env!("CARGO_BIN_EXE_janus_interop_aggregator"));
+    helper_command.arg("--port").arg(format!("{}", helper_port));
+    helper_command.arg("--postgres-url").arg(format!(
+        "postgres://postgres@127.0.0.1:{}/postgres",
+        helper_postgres_port
+    ));
+    let mut collector_command = Command::new(env!("CARGO_BIN_EXE_janus_interop_collector"));
+    collector_command
+        .arg("--port")
+        .arg(format!("{}", collector_port));
+    let commands = [
+        client_command,
+        leader_command,
+        helper_command,
+        collector_command,
+    ];
+    let mut drop_guards = Vec::with_capacity(commands.len());
+    for mut command in commands {
+        let mut drop_guard = ChildProcessCleanupDropGuard(
+            command
+                .stdout(Stdio::piped())
+                .stderr(Stdio::piped())
+                .spawn()?,
+        );
+        tokio::spawn(forward_stdout(ChildStdout::from_std(
+            drop_guard.0.stdout.take().unwrap(),
+        )?));
+        tokio::spawn(forward_stderr(ChildStderr::from_std(
+            drop_guard.0.stderr.take().unwrap(),
+        )?));
+        drop_guards.push(drop_guard);
+    }
+
+    // Step 3: Try opening a TCP connection to each container's port, and
+    // retry until it succeeds.
+    for port in [client_port, leader_port, helper_port, collector_port] {
+        wait_for_tcp_server(port).await?;
```

Comment:
I don't think we can assume for all DAP implementations that the TCP port being bound necessarily means the server is ready for requests. I think you'll need to do some retries in the HTTP client accessing `/internal/test/endpoint_for_task` anyway, at which point you might as well not bother polling the TCP port.

Reply:
This test is already making assumptions that it's just talking to Janus's interop test servers (i.e., it hardcodes the aggregator URLs assuming that `endpoint_for_task` returns "/") but I'll open an issue to track this more generally on the design document repository.